proposals/mmap_proposal.md

```diff
@@ -0,0 +1,342 @@
+
+- It is efficient for random access on large files.
+- It allows multiple processes to share memory.
+- It can improve performance by reducing system calls and copy operations.
+
+It uses demand paging, which means that files are not immediately read from disk.
+The OS manages this mapping and returns additional pages of data as needed.
+
+## Proposal
+
+The introduction of a `mmap` module, with two primary objects:
+
+A `MmapMode`, an enum struct, identifying the four initial `mmap` configurations:
+
+- Read Only
+- Write Only
+- Exec
+- Copy On Write
+
+A `Mmap` object, responsible for creating and working with mmap files.
+
+### High Level Usage
+
+An initial implementation of `Mmap` in a read-only context can be used as follows.
+Note, the API will provide functionality similar to the existing `FileHandle` API.
+
+```mojo
+    var length: Int32 = 24;
+    var fd: FileDescriptor = open("./test.txt", "r");
+
+    # Initialize a mmap object for the particular file, with read and shared
+    # permissions
+    var mmap = Mmap(fd=fd, length=length, offset=0, mode=MmapMode.READ)
+
+    # Print out the entire file as a String
+    print(String(mmap.read_bytes()))
+
+    # Memory mapped files must be closed once they are finished
+    mmap.close()
+```
+
+Whereas an initial implementation of `Mmap` in a write context can be used as follows:
+
+```mojo
+    var length: Int32 = 24;
+    var fd: FileDescriptor = open("./test.txt", "r+")
+
+    # Initialize a mmap object for the particular file, with read/write and shared permissions
+    var mmap = Mmap(fd=fd, length=length, offset=0, mode=MmapMode.WRITE)
+
+    # Print out the entire file as a String
+    print(String(mmap.read_bytes()))
+
+    # Write bytes to mmap
+    mmap.write("hello world")
+
+    # Memory mapped files must be closed once they are finished
+    mmap.close()
+```
+
+### Initial API
+
+An initial API for `MmapMode` and `Mmap` could look like the below.
+
+```mojo
+    struct MmapMode:
+        alias READ: String = "READ"
+        alias WRITE: String = "WRITE"
+        alias EXEC: String = "EXEC"
+
+    struct Mmap:
+        var pointer: UnsafePointer[UInt8];
+        var length: Int32;
+        var offset: Int32;
+        var mode: String;
+
+    @staticmethod
+    fn _get_prot(mode: String) raises -> Int32:
+        """Helper method to translate between `MmapMode` and OS accepted prot values"""
+        ...
+
+    @staticmethod
+    fn _get_flags(mode: String) raises -> Int32:
+        """Helper method to translate between `MmapMode` and OS accepted flags values"""
+        ...
+
+    fn __init__(inout self, fd: FileDescriptor, length: Int32, offset: Int32, mode: String, address: UnsafePointer[UInt8]) raises:
+        """Given a FileDescriptor, an expected length and offset, mmap configuration, and an initial address, initialize a `mmap` object"""
+        ...
+
+    fn __init__(inout self, length: Int32, offset: Int32, mode: String, address: UnsafePointer[UInt8]) raises:
```

Comment:
```suggestion
    fn __init__(inout self, length: Int32, offset: Int32, mode: MmapMode, address: UnsafePointer[UInt8]) raises:
```

**Suggestion** Since we have the `MmapMode` struct, should we use it here (and elsewhere) instead of having this argument be "stringly typed"? 🙂